### ISSUE TITLE
Add a check to detect mutation of the inputs during backward

### ISSUE THREAD

**Issue by soulitzer** [pytorch/pytorch#81417]
Modifying the grad_input during the backward pass is unexpected to the user and can produce silently incorrect gradients. However, currently you're allowed to do something like this:
```
import torch

class Func(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        return x

    @staticmethod
    def vjp(ctx, x):
        x += 1
        return x

a = torch.rand(2, 4, requires_grad=True)

out = Func.apply(a)
grad_out = torch.ones_like(out)
grad_a, = torch.autograd.grad(out, (a,), grad_outputs=grad_out)

print(grad_out) 
```
